In [10]:
from apyori import apriori
import pyfpgrowth
import numpy as np
import pandas as pd
import csv
from matplotlib import pyplot as plt

In [11]:
f = open('data/data3.txt','r')

In [12]:
transaction = []

In [13]:
for i, line in enumerate(f):
    print(line.strip('\n'))
    text = line.strip('\n')
    array = text.split(',')
    array = list(map(int,array))
    transaction.append(array)

2514,4089
1898,1899,1900,4126,4127,4991,5976
1844,4079
2514,4089
1374,1942,2472,4150
1943,4025,4150,4879
1374,1942,1943,4150,4879
1374,1942,2472,4150
1844,4079
1850,4090
2388,4078
1955,1968,2388,4030,4078
1951,1955,1968,2388,4027,4030,4078,4880,5859
1906,2391,4084,4869
1906,1907,2391,4084,4869
2393,2396,4079,4098,5802
1941,2018,2475,4029,4038,4152,4173,4193,4938,4939,4940,5785
1855,2434,4091
1855,2434,4091,4097,5785
1855,4091
1374,1942,1973,2472,2494,2612,4025,4033,4150,4155,4161,5785,5873,5970
1942,2472,4025,4150
2514,4089
1374,1942,1943,2472,4025,4150,4879,5853
2514,4089
1374,1942,1943,2472,4025,4150,4879,5853
1374,1942,2472,4025,4150,4155,5785
1942,2472,4150
1942,1943,2472,4025,4150,4879,5853
2514,4089
1943,2472,4150
2514,4089
1374,1942,2472,4150
1928,2463,2476,4080,4142,4961,4962,5837,6781
1838,1839,1928,2463,2476,4080,4142,4961,4962,5837,6781
1838,1928,2476,2603,4142,4961,4962,5837
1928,1929,2476,2603,4142,4961,4962,5837
2393,4079,5802
1881,2059,2060,2061,2411,2546,2550,2584,2585,

ValueError: invalid literal for int() with base 10: ''

In [14]:
transaction

[[2514, 4089],
 [1898, 1899, 1900, 4126, 4127, 4991, 5976],
 [1844, 4079],
 [2514, 4089],
 [1374, 1942, 2472, 4150],
 [1943, 4025, 4150, 4879],
 [1374, 1942, 1943, 4150, 4879],
 [1374, 1942, 2472, 4150],
 [1844, 4079],
 [1850, 4090],
 [2388, 4078],
 [1955, 1968, 2388, 4030, 4078],
 [1951, 1955, 1968, 2388, 4027, 4030, 4078, 4880, 5859],
 [1906, 2391, 4084, 4869],
 [1906, 1907, 2391, 4084, 4869],
 [2393, 2396, 4079, 4098, 5802],
 [1941, 2018, 2475, 4029, 4038, 4152, 4173, 4193, 4938, 4939, 4940, 5785],
 [1855, 2434, 4091],
 [1855, 2434, 4091, 4097, 5785],
 [1855, 4091],
 [1374,
  1942,
  1973,
  2472,
  2494,
  2612,
  4025,
  4033,
  4150,
  4155,
  4161,
  5785,
  5873,
  5970],
 [1942, 2472, 4025, 4150],
 [2514, 4089],
 [1374, 1942, 1943, 2472, 4025, 4150, 4879, 5853],
 [2514, 4089],
 [1374, 1942, 1943, 2472, 4025, 4150, 4879, 5853],
 [1374, 1942, 2472, 4025, 4150, 4155, 5785],
 [1942, 2472, 4150],
 [1942, 1943, 2472, 4025, 4150, 4879, 5853],
 [2514, 4089],
 [1943, 2472, 4150],
 [251

# apriori

In [17]:
result = list(apriori(transaction))
print(result)

[RelationRecord(items=frozenset({1942}), support=0.14285714285714285, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({1942}), confidence=0.14285714285714285, lift=1.0)]), RelationRecord(items=frozenset({1943}), support=0.12337662337662338, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({1943}), confidence=0.12337662337662338, lift=1.0)]), RelationRecord(items=frozenset({2391}), support=0.12337662337662338, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({2391}), confidence=0.12337662337662338, lift=1.0)]), RelationRecord(items=frozenset({2395}), support=0.11038961038961038, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({2395}), confidence=0.11038961038961038, lift=1.0)]), RelationRecord(items=frozenset({2472}), support=0.18831168831168832, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({2472}), confidence=0.18831168831

# FPgrowth

In [18]:
patterns = pyfpgrowth.find_frequent_patterns(transaction,2)
rules = pyfpgrowth.generate_association_rules(patterns,0.02)

In [19]:
rules

{(1844,): ((4079,), 1.0),
 (4079,): ((2393,), 0.75),
 (2393,): ((4079,), 1.0),
 (5802,): ((2393, 4079), 1.0),
 (2393, 4079): ((5802,), 0.3333333333333333),
 (2393, 5802): ((4079,), 1.0),
 (4079, 5802): ((2393,), 1.0),
 (1973,): ((2472, 2494, 4025, 4161, 5970), 1.0),
 (2472,): ((4150,), 0.9655172413793104),
 (1973, 2472): ((2494, 4025, 4161, 5970), 1.0),
 (1973, 4025): ((2472, 2494, 4161, 5970), 1.0),
 (2472, 4025): ((4150,), 0.9333333333333333),
 (2494,): ((1973, 2472, 4025, 4161, 5970), 1.0),
 (1973, 2494): ((2472, 4025, 4161, 5970), 1.0),
 (2494, 4025): ((1973, 2472, 4161, 5970), 1.0),
 (2472, 2494): ((1973, 4025, 4161, 5970), 1.0),
 (1973, 2472, 2494): ((4025, 4161, 5970), 1.0),
 (1973, 2472, 4025): ((2494, 4161, 5970), 1.0),
 (1973, 2494, 4025): ((2472, 4161, 5970), 1.0),
 (2472, 2494, 4025): ((1973, 4161, 5970), 1.0),
 (4161,): ((1973, 2472, 2494, 4025, 5970), 1.0),
 (1973, 4161): ((2472, 2494, 4025, 5970), 1.0),
 (2494, 4161): ((1973, 2472, 4025, 5970), 1.0),
 (4025, 4161): ((197

In [24]:
patterns2 = pyfpgrowth.find_frequent_patterns(transaction,8)
rules2 = pyfpgrowth.generate_association_rules(patterns,0.001)

In [25]:
rules2

{(1844,): ((4079,), 1.0),
 (4079,): ((2393,), 0.75),
 (2393,): ((4079,), 1.0),
 (5802,): ((2393, 4079), 1.0),
 (2393, 4079): ((5802,), 0.3333333333333333),
 (2393, 5802): ((4079,), 1.0),
 (4079, 5802): ((2393,), 1.0),
 (1973,): ((2472, 2494, 4025, 4161, 5970), 1.0),
 (2472,): ((4150,), 0.9655172413793104),
 (1973, 2472): ((2494, 4025, 4161, 5970), 1.0),
 (1973, 4025): ((2472, 2494, 4161, 5970), 1.0),
 (2472, 4025): ((4150,), 0.9333333333333333),
 (2494,): ((1973, 2472, 4025, 4161, 5970), 1.0),
 (1973, 2494): ((2472, 4025, 4161, 5970), 1.0),
 (2494, 4025): ((1973, 2472, 4161, 5970), 1.0),
 (2472, 2494): ((1973, 4025, 4161, 5970), 1.0),
 (1973, 2472, 2494): ((4025, 4161, 5970), 1.0),
 (1973, 2472, 4025): ((2494, 4161, 5970), 1.0),
 (1973, 2494, 4025): ((2472, 4161, 5970), 1.0),
 (2472, 2494, 4025): ((1973, 4161, 5970), 1.0),
 (4161,): ((1973, 2472, 2494, 4025, 5970), 1.0),
 (1973, 4161): ((2472, 2494, 4025, 5970), 1.0),
 (2494, 4161): ((1973, 2472, 4025, 5970), 1.0),
 (4025, 4161): ((197

# 결과 비교
## Apriori
|item1|item2|minsupport|confidence|lift|
|-|-|----------|----------|----|
|1838|2391|0.10878489|0.57235421|3.56586921|
|2391|1838|0.10878489|0.67774936|3.56586921|
|1838|4080|0.18513957|0.97408207|4.44356541|
|4080|1838|0.18513957|0.84456929|4.44356541|
|1844|4079|0.10016420|0.96442688|7.14086284|
|4079|1844|0.10016420|0.74164134|7.14086284|
|2391|4080|0.12848933|0.80051151|3.65177160|
|4080|2391|0.12848933|0.58614232|3.65177160|
|2472|4150|0.13998358|0.97428571|6.06997442|
|4150|2472|0.13998358|0.87212276|6.06997442|

## FPgrowth
* 엄청 많음
* 따로 찾아서 표현할 것